src/core/thread/mle_router.cpp

```diff
@@ -4483,6 +4484,59 @@ Error MleRouter::SendTimeSync(void)
 }
 #endif // OPENTHREAD_CONFIG_TIME_SYNC_ENABLE
 
+Error MleRouter::DetachGracefully(otDetachGracefullyCallback aCallback, void *aContext)
+{
+    Error error = kErrorNone;
+
+    if (IsChild())
+    {
+        ExitNow(error = Mle::DetachGracefully(aCallback, aContext));
+    }
+
+    VerifyOrExit(!IsDetachingGracefully(), error = kErrorBusy);
+
+    OT_ASSERT(mDetachGracefullyCallback == nullptr);
+
+    mDetachGracefullyCallback = aCallback;
+    mDetachGracefullyContext  = aContext;
+
+    mDetachGracefullyTimer.Start(kDetachGracefullyTimeout);
+
+    if (IsRouter())
+    {
+        SendAddressRelease(&MleRouter::HandleDetachGracefullyAddressReleaseResponse, this);
+    }
+    else
+    {
+        mStopThreadTask.Post();
+    }
+
+exit:
+    return error;
+}
+
+void MleRouter::HandleDetachGracefullyAddressReleaseResponse(void *               aContext,
+                                                             otMessage *          aMessage,
+                                                             const otMessageInfo *aMessageInfo,
+                                                             Error                aResult)
+{
+    static_cast<MleRouter *>(aContext)->HandleDetachGracefullyAddressReleaseResponse(aMessage, aMessageInfo, aResult);
+}
+
+void MleRouter::HandleDetachGracefullyAddressReleaseResponse(otMessage *          aMessage,
```

Comment:
We can consider not passing the parameters from the `static` version of this method?

Reply:
Done.